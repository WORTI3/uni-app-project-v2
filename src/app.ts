// Dependencies
import express, { ErrorRequestHandler } from 'express';
import nunjucks from 'nunjucks';
import createError from 'http-errors';
import path from 'path';
import cookieParser from 'cookie-parser';
import session from 'express-session';
import csrf from 'csurf';
import sqlite3 from 'sqlite3';
import sqliteStoreFactory from 'express-session-sqlite';
// App routers
import { initPassport } from './middleware/passport';
import { homeRouter } from './routes/home';
import { authRouter } from './routes/auth';
import { indexRouter } from './routes/index';
import { Session } from './types';

const app = express();
// Db
const SqliteStore = sqliteStoreFactory(session);
const store = new SqliteStore({
  driver: sqlite3.Database,
  path: './src/database/sessions.db', // Specify the path to the SQLite database file
  ttl: 86400, // Session Time To Live in seconds
  cleanupInterval: 300000, // Cleanup interval in milliseconds for deleting expired sessions
});

// Nunjucks configuration
nunjucks.configure(['./src/views', './src/_layouts'], {
  autoescape: true,
  express: app,
});
// View engine setup
app.set('view engine', 'njk');
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());

// Static assets (main.css)
app.use(express.static(path.join('./', 'public')));

// Adds CSRF protection to the application.
app.use(csrf({ cookie: true }));

// Init session
app.use(
  session({
    secret: 'nobody knows',
    resave: false,
    saveUninitialized: false,
    store,
  }),
);
// passport middleware
initPassport(app);

/**
 * Sets messages and tone to locals from the current session.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 */
app.use((req, res, next) => {
  const session = req.session as Session;

  if (!session.errorFields) {
    // when no session is present
    const inputValues = {
      username: req.body.username ?? '',
      password: '', // dont want to pass password value through here
    };
    const fields = [];
    for (const [key, value] of Object.entries(inputValues)) {
      fields.push({
        field: key,
        value: value,
      });
    }
    session.errorFields = fields;
  }

  // todo remove messages and use errorFields.error

  if (session.messages || session.errorFields) {
    console.log('error fields:', session.errorFields);

    const msgs = session?.messages ?? [];
    const tone = session.msgTone ?? null;
    res.locals.messages = msgs;
    res.locals.msgTone = tone;
    res.locals.hasMessages =
      Boolean(msgs.length) || Boolean(session.errorFields);
    res.locals.errorFields = session.errorFields;
    session.messages = [];
    // session.errorFields = [];
    session.msgTone = undefined;
  }
  next();
});

/**
 * Middleware function that adds a CSRF token to the response locals object.
 * If the environment is set to "test", a hardcoded token is used instead.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
app.use((req, res, next) => {
  res.locals.csrfToken = req.csrfToken();
  next();
});

// App routes
app.use('/', homeRouter);
app.use('/', indexRouter);
app.use('/', authRouter);

// Catch 404 and forward to error handler
app.use((_req, _res, next) => {
  next(createError(404));
});

// Error handler
const errorHandler: ErrorRequestHandler = (err, req, res, _next) => {
  // Set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};
  res.status(err.status || 500).render('error');
};

app.use(errorHandler);

export default app;
