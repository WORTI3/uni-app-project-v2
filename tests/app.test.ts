import request from 'supertest';
import cheerio from 'cheerio';
import app, { errorHandler } from '../src/app';
import supertest from 'supertest';
import express from 'express';

describe('Nunjucks configuration', () => {
  it("should set the view engine to 'njk'", () => {
    expect(app.get('view engine')).toBe('njk');
  });
});

describe('handle 404 errors', () => {
  test('GET /non-existent-route should return 404', async () => {
    await request(app).get('/non-existent-route').expect(404);
  });

  it('should render error template with status 404 for not found error', async () => {
    // given / when
    const result = await request(app).get('/invalid_route').expect(404);
    const $ = cheerio.load(result.text);

    // then
    expect(result.status).toBe(404);
    expect($('h1').text()).toBeDefined();
    expect($('h2').text()).toBe('Error: Not Found');
    expect($('p')).toHaveLength(2);
    expect($('p').text()).toBe(
      'Looks like the application has thrown an errorPlease try again later',
    );
    expect(result.text).toMatchSnapshot();
  });
});

describe('unit tests for app.js routers', () => {
  // '/' for home is the same as this test
  test('GET / should return 200', async () => {
    const body = {
      groupName: 'Test Group Name',
      email: 'test@example.com',
      country: 'US',
    };
    await supertest(app).get('/').send(body).expect(200);
  });

  test('GET /login should return 200', async () => {
    await request(app).get('/login').expect(200);
  });

  test('GET /signup should return 200', async () => {
    await request(app).get('/signup').expect(200);
  });
});

describe('errorHandler test errors', () => {
  const tempApp = express();
  tempApp.use(errorHandler);
  const errorMessage = 'Test error message';
  const itif = (env?: string) => (env === 'test' ? it : it.skip);

  itif(process.env.NODE_ENV)(
    'should respond with error message and status 500',
    async () => {
      // given
      tempApp.use((_req, _res, next) => {
        next(new Error(errorMessage));
      });

      // when
      const response = await request(tempApp).get('/');

      // then
      const $ = cheerio.load(response.text);
      expect($('title').text()).toBe('Error');
      expect($('pre').text()).toContain('Error: Test error message');
      expect(response.status).toBe(500);
    },
  );

  itif(process.env.NODE_ENV)(
    'should set locals message and error in development environment',
    async () => {
      // given

      tempApp.use((_req, _res, next) => {
        next(new Error(errorMessage));
      });

      // when
      const response = await request(tempApp).get('/');

      // then
      expect(response.text).toContain(errorMessage);
      const $ = cheerio.load(response.text);
      expect($('title').text()).toBe('Error');
      expect($('pre').text()).toContain('Error: Test error message');
      expect(response.status).toBe(500);
    },
  );
});

process.env.NODE_ENV = 'production';

describe('errorHandler production errors', () => {
  const tempApp = express();
  tempApp.use(errorHandler);
  const errorMessage = 'Test error message';
  const itif = (env?: string) => (env === 'production' ? it : it.skip);

  itif(process.env.NODE_ENV)(
    'should respond with error message and status 500',
    async () => {
      // given
      tempApp.use((_req, _res, next) => {
        next(new Error(errorMessage));
      });

      // when
      const response = await request(tempApp).get('/');

      // then
      const $ = cheerio.load(response.text);
      expect($('title').text()).toBe('Error');
      expect($('pre').text()).toBe('Internal Server Error');
      expect(response.status).toBe(500);
    },
  );

  itif(process.env.NODE_ENV)(
    'should not set locals error in production environment',
    async () => {
      // given
      tempApp.use((_req, _res, next) => {
        next(new Error(errorMessage));
      });

      // when
      const response = await request(tempApp).get('/');

      // then
      expect(response.status).toBe(500); // Ensure status 500 for production
      expect(response.text).not.toContain(errorMessage); // Error message should not be exposed in production

      const $ = cheerio.load(response.text);
      expect($('title').text()).toBe('Error');
      expect($('pre').text()).not.toContain('Error: Test error message');
    },
  );
});
