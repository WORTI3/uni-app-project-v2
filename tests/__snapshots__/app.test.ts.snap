// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errorHandler production errors should not set locals error in production environment 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Internal Server Error</pre>
</body>
</html>
"
`;

exports[`errorHandler production errors should respond with error message and status 500 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Internal Server Error</pre>
</body>
</html>
"
`;

exports[`errorHandler test errors should respond with error message and status 500 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Error: Test error message<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/tests/app.test.ts:66:14<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:91:12)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at expressInit (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/middleware/init.js:40:5)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at query (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/middleware/query.js:45:5)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at Function.handle (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:175:3)<br> &nbsp; &nbsp;at Function.handle (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/application.js:181:10)<br> &nbsp; &nbsp;at Server.app (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/express.js:39:9)<br> &nbsp; &nbsp;at Server.emit (node:events:514:28)<br> &nbsp; &nbsp;at parserOnIncoming (node:_http_server:1107:12)<br> &nbsp; &nbsp;at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)</pre>
</body>
</html>
"
`;

exports[`errorHandler test errors should set locals message and error in development environment 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Error: Test error message<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/tests/app.test.ts:66:14<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:91:12)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at expressInit (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/middleware/init.js:40:5)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at query (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/middleware/query.js:45:5)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/layer.js:95:5)<br> &nbsp; &nbsp;at trim_prefix (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:328:13)<br> &nbsp; &nbsp;at /Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:286:9<br> &nbsp; &nbsp;at Function.process_params (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:346:12)<br> &nbsp; &nbsp;at next (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:280:10)<br> &nbsp; &nbsp;at Function.handle (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/router/index.js:175:3)<br> &nbsp; &nbsp;at Function.handle (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/application.js:181:10)<br> &nbsp; &nbsp;at Server.app (/Users/arcwo/projects/uni-app-project-v2/node_modules/express/lib/express.js:39:9)<br> &nbsp; &nbsp;at Server.emit (node:events:514:28)<br> &nbsp; &nbsp;at parserOnIncoming (node:_http_server:1107:12)<br> &nbsp; &nbsp;at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)</pre>
</body>
</html>
"
`;

exports[`handle 404 errors should render error template with status 404 for not found error 1`] = `
"
<!DOCTYPE html>
<html lang="en">
    
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="description" content="Faultifier - a ticket handling admin service.">
        <meta name="theme-color" content="blue">
		<title> - Faultifier</title>
		<link rel="stylesheet" type="text/css" href="/main.css">
	</head>

    <div class="flex flex-col h-screen justify-between overflow-hidden">
        <nav class="flex items-center justify-between flex-wrap bg-gray-100 p-6">
    <div class="flex items-center flex-shrink-0 text-picton-blue-600 mr-6">
        <a href="/"><span class="font-bold text-xl tracking-tight font-sans">FAULTIFIER</span></a>
    </div>
    
    
</nav>

        <body class="bg-gradient-to-tr from-picton-blue-600 to-yellow-500 overflow-hidden">
            <main>

<div class="p-12 bg-white sm:mx-auto rounded-md w-1/2 max-sm:w-full max-sm:p-6">
  
  <h1 class="text-center">404</h1>
  <h2>Error: Not Found</h2>
  <p>Looks like the application has thrown an error</p>
  <p>Please try again later</p>
</div>
</main>
        </body>

        
        <footer class="lg:text-left bg-gray-100 text-gray-600">
            <div class="text-center p-6">
                <span>Â© 2023 </span>
                <a class="text-gray-600 font-semibold" href="https://github.com/WORTI3">Wortie</a>
            </div>
        </footer>
    </div>
</html>"
`;
